### PR TITLE
feat: refactor entry

### PR DESCRIPTION
**Pull request by caohuilin**
## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by zllkjc** [comment]
all the `addRuntimeExports` should comment deprecated ?

**Comment by zllkjc** [comment]
what's the different between the runtime/core/server and runtime/ssr

File: packages/runtime/plugin-garfish/src/cli/index.ts [hunk 1]
```diff
@@ -194,123 +195,27 @@ export const garfishPlugin = ({
...
-          imports.push({
-            value: runtimePluginName,
-            specifiers: [
-              {
-                imported: 'masterApp',
-              },
-            ],
-          });
+      async generateEntryCode({ entrypoints }) {
+        const resolveOptions = api.useResolvedConfigContext();
```
Comment (zllkjc) [inline packages/runtime/plugin-garfish/src/cli/index.ts:212]: so, garfishPlugin must run after runtimePlugin..

File: packages/runtime/plugin-garfish/src/cli/code.ts
Comment (zllkjc) [inline packages/runtime/plugin-garfish/src/cli/code.ts:1]: same problem as runtime cli code.ts

File: packages/runtime/plugin-garfish/src/cli/template.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  srcDirectory: string;
+  internalSrcAlias: string;
+  metaName: string;
+  entry: string;
+  customEntry?: string | false;
+  mountId?: string;
+}) =>
+  customEntry
+    ? `import '${entry.replace(srcDirectory, internalSrcAlias)}'
+export * from '${entry.replace(srcDirectory, internalSrcAlias)}'`
```
Comment (zllkjc) [inline packages/runtime/plugin-garfish/src/cli/template.ts:18]: if we already export `provider` in line 29, do we also need to `export *` here ?

File: packages/runtime/plugin-runtime/src/cli/code.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          const indexServerFile = path.resolve(
+            internalDirectory,
+            `./${entryName}/${SERVER_ENTRY_POINT_FILE_NAME}`,
+          );
+
+          fs.outputFileSync(indexServerFile, indexServerCode, 'utf8');
+        }
+
+        // register.js
+        const registerCode = template.register();
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/runtime/plugin-runtime/src/cli/code.ts:129]: we need this file for garfish and vmock register logic

File: packages/runtime/plugin-runtime/src/core/client/index.tsx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        context.initialData = initialData;
+      }
+      const rootElement =
+        id && typeof id !== 'string'
+          ? id
+          : document.getElementById(id || 'root')!;
+
+      async function ModernRender(App: React.ReactElement) {
+        const renderFunc = IS_REACT18 ? renderWithReact18 : renderWithReact17;
+        return renderFunc(React.cloneElement(App, { context }), rootElement);
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/runtime/plugin-runtime/src/core/client/index.tsx:75]: image
App need this context

File: packages/runtime/plugin-runtime/src/core/react/index.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import { getGlobalRunner } from '../plugin/runner';
+
+export function createRoot(
+  UserApp?: React.ComponentType | null,
+  config?: { router: { basename: string } },
+) {
+  const App = UserApp || getGlobalApp();
+  const WrapperComponent: React.ComponentType<any> = props => {
+    return React.createElement(
+      App || React.Fragment,
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/runtime/plugin-runtime/src/core/react/index.tsx:12]: routes entry is no App, this is anticipated

File: packages/runtime/plugin-runtime/src/router/runtime/plugin.tsx [hunk 6]
```diff
@@ -88,7 +87,9 @@ export const routerPlugin = ({
                 window._SERVER_DATA?.router.baseUrl ||
                 select(location.pathname);
               const _basename =
-                baseUrl === '/' ? urlJoin(baseUrl, basename) : baseUrl;
+                baseUrl === '/'
+                  ? urlJoin(baseUrl, config?.router?.basename || basename)
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/runtime/plugin-runtime/src/router/runtime/plugin.tsx:91]: image

originalBaseUrl is the highest priority

File: packages/runtime/plugin-garfish/src/cli/index.ts (hunk 1 above)
Comment (caohuilin) [reply packages/runtime/plugin-garfish/src/cli/index.ts:212]: yes

File: packages/runtime/plugin-garfish/src/cli/template.ts (hunk 2 above)
Comment (caohuilin) [reply packages/runtime/plugin-garfish/src/cli/template.ts:18]: when use customEntry, exports provider need user to export?

File: packages/solutions/app-tools/src/types/hooks.ts [hunk 7]
```diff
@@ -22,41 +22,16 @@ export interface ImportSpecifier {
...
-  specifiers: ImportSpecifier[];
-  value: string;
-  initialize?: string;
-}
-
 export interface RuntimePlugin {
   name: string;
-  options: string;
-  args?: string;
+  implementation: string;
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/solutions/app-tools/src/types/hooks.ts:27]: use path

File: packages/runtime/plugin-runtime/src/core/client/index.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/runtime/plugin-runtime/src/core/client/index.tsx]: 待讨论

File: packages/runtime/plugin-garfish/src/cli/code.ts
Comment (caohuilin) [reply packages/runtime/plugin-garfish/src/cli/code.ts:1]: async entry

File: packages/toolkit/runtime-utils/src/universal/index.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export * from './request';
```
[the remark this answers is not in the document]
Comment (GiveMe-A-Name) [reply packages/toolkit/runtime-utils/src/universal/index.ts:1]: Got it

File: packages/runtime/plugin-runtime/src/router/runtime/plugin.node.tsx [hunk 9]
```diff
@@ -59,6 +41,11 @@ export function createFetchHeaders(
   return headers;
 }
 
+// eslint-disable-next-line import/no-mutable-exports
+export let finalRouteConfig: RouterConfig['routesConfig'] = {
```
Comment (yimingjfe) [inline packages/runtime/plugin-runtime/src/router/runtime/plugin.node.tsx:45]: Why we can not put `finalRouteConfig` in the setup function?

File: packages/runtime/plugin-runtime/src/router/runtime/plugin.node.tsx [hunk 10]
```diff
@@ -144,47 +135,40 @@ export const routerPlugin = ({
           context.routerContext = routerContext;
           context.routes = routes;
           // set routeManifest in context to be consistent with csr context
-          context.routeManifest = context.ssrContext!
-            .routeManifest as RouteManifest;
+          // context.routeManifest = context.ssrContext!
```
Comment (yimingjfe) [inline packages/runtime/plugin-runtime/src/router/runtime/plugin.node.tsx:138]: It seems that there are some comments left here

Comment (GiveMe-A-Name) [reply packages/runtime/plugin-runtime/src/router/runtime/plugin.node.tsx:138]: Got it

**Comment by GiveMe-A-Name** [comment]
> what's the different between the runtime/core/server and runtime/ssr

runtime/core/server will replace of runtime/ssr

File: packages/runtime/plugin-runtime/src/core/server/stream/createReadableStream.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            ssrConfig,
+            renderLevel,
+            runtimeContext,
+            config,
+            styledComponentsStyleTags,
+          }).then(({ shellAfter, shellBefore }) => {
+            const body = new Transform({
+              transform(chunk, _encoding, callback) {
+                try {
+                  if (shellChunkStatus !== ShellChunkStatus.FINIESH) {
```
Comment (zllkjc) [inline packages/runtime/plugin-runtime/src/core/server/stream/createReadableStream.ts:56]: FINIESH -> FINISH ?

File: packages/solutions/app-tools/src/commands/serve.ts [hunk 12]
```diff
@@ -56,7 +56,7 @@ export const start = async (api: PluginAPI<AppTools<'shared'>>) => {
       output: {
         path: userConfig.output.distPath?.root,
         ...(userConfig.output || {}),
-      },
+      } as any,
```
[the remark this answers is not in the document]
Comment (GiveMe-A-Name) [reply packages/solutions/app-tools/src/commands/serve.ts:59]: Got it

**Comment by zllkjc** [comment]
Now `context` is a reserved field of <ModernRoot>, can we use `_context` or more private fields?

File: packages/runtime/plugin-runtime/src/core/server/stream/createReadableStream.ts (hunk 11 above)
Comment (GiveMe-A-Name) [reply packages/runtime/plugin-runtime/src/core/server/stream/createReadableStream.ts:56]: Got it

File: packages/solutions/app-tools/src/plugins/analyze/getFileSystemEntry.ts [hunk 13]
```diff
@@ -62,14 +84,28 @@ const scanDir = (
...
         return {
           entryName,
           isMainEntry: false,
-          entry: entryFile,
+          entry: enableCustomEntry ? customEntryFile || entryFile : entryFile,
+          customServerEntry,
           absoluteEntryDir: path.resolve(dir),
           isAutoMount: true,
           customBootstrap,
+          customEntry: enableCustomEntry ? customEntryFile : false,
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/solutions/app-tools/src/plugins/analyze/getFileSystemEntry.ts:97]: customEntry  用 boolean，template 用 entry

File: packages/server/core/src/types/config/output.ts
[the remark this answers is not in the document]
Comment (GiveMe-A-Name) [reply packages/server/core/src/types/config/output.ts:1]: Before this PR, there outputConfig will pass by in build entry. Like this 
```ts
createApp(App, [
  ssr({ enableInlineScripts: true, enableInlineStyles: false })
]);
```
But now, there are not exist ssr plugin.
So this config will pass by server-core

File: packages/runtime/plugin-router-v5/src/runtime/plugin.tsx [hunk 14]
```diff
@@ -130,13 +147,19 @@ export const routerPlugin = ({
...
+                 * when exist createRoutes function, App.tsx must be exist, and support Component props
+                 * this is compatible config routes
+                 */
                 return (
                   <Router history={history}>
                     {createRoutes ? (
-                      <App {...props} Component={createRoutes()} />
+                      <App Component={createRoutes()} />
+                    ) : App && !finalRouteConfig?.routes ? (
+                      <App {...props} />
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/runtime/plugin-router-v5/src/runtime/plugin.tsx:156]: for v5 + App.tsx

File: packages/runtime/plugin-runtime/src/cli/template.ts [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+  metaName: string;
+  entry: string;
+  customEntry?: string | false;
+}) => {
+  return `import { setGlobalContext } from '@${metaName}/runtime/context'
+
+import App from '${
+    customEntry
+      ? entry
+          .replace('entry.tsx', 'App.tsx')
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/runtime/plugin-runtime/src/cli/template.ts:159]: comment 一下这里的 entry

File: packages/runtime/plugin-runtime/src/core/context/index.ts [hunk 16]
```diff
@@ -1 +1,70 @@
...
+
+export function getGlobalRoutes(): undefined | (NestedRoute | PageRoute)[] {
+  return globalContext.routes;
+}
+
+export function getGlobalAppInit() {
+  return (
+    globalContext.appInit ||
+    (getGlobalApp() as any)?.init ||
+    (getGlobalLayoutApp() as any)?.init
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/runtime/plugin-runtime/src/core/context/index.ts:56]: getGlobalLayoutApp 是获取的 pages 路由 _app 的 App

File: packages/runtime/plugin-garfish/src/runtime/utils.ts [hunk 17]
```diff
@@ -0,0 +1,13 @@
+declare const __GARFISH_EXPORTS__: string;
+
+export function isRenderGarfish(params?: { appName?: string }) {
+  const renderByGarfish =
+    typeof __GARFISH_EXPORTS__ !== 'undefined' ||
+    (typeof window !== 'undefined' &&
+      window.Garfish &&
+      window.Garfish.activeApps &&
+      window.Garfish.activeApps.some(
+        app => app.appInfo?.name === params?.appName,
```
Comment (nyqykk) [inline packages/runtime/plugin-garfish/src/runtime/utils.ts:10]: i checked the build products, the 'params' has not been pass to 'isRenderGarfish', do this mean the  ' app.appInfo?.name === params?.appName,' will always be false?
image

Comment (caohuilin) [reply packages/runtime/plugin-garfish/src/runtime/utils.ts:10]: Currently, yes

**Review by KyrieLii — APPROVED** [review]

**Pull request merged**